### PR TITLE
LUCENE-9766: Hunspell: add API for retrieving dictionary morphologica…

### PR DESCRIPTION
**Pull request by donnerpeter**
…l data and stemming

# Description

We need to rank suggestions based on metadata associated with corresponding dictionary entries. For that, we need a stemming API (to get the entries) and an API to get the morphological data (where we'd store the needed info)

# Solution

Add public `Hunspell.getRoots` and `Dictionary.lookupEntries`

# Tests

Yes, a test for each new introduced method

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: lucene/analysis/common/src/test/org/apache/lucene/analysis/hunspell/StemmerTestBase.java [hunk 1]
```diff
@@ -43,6 +43,11 @@ static void init(String affix, String dictionary) throws IOException, ParseExcep
 
   static void init(boolean ignoreCase, String affix, String... dictionaries)
       throws IOException, ParseException {
+    stemmer = new Stemmer(loadDictionary(ignoreCase, affix, dictionaries));
+  }
+
+  static Dictionary loadDictionary(boolean ignoreCase, String affix, String... dictionaries)
```
Comment (donnerpeter) [inline lucene/analysis/common/src/test/org/apache/lucene/analysis/hunspell/StemmerTestBase.java:49]: extract a method to call from a test

File: lucene/analysis/common/src/test/org/apache/lucene/analysis/hunspell/TestAllDictionaries.java [hunk 2]
```diff
@@ -205,6 +205,7 @@ private static String memoryUsageSummary(Dictionary dic) {
         + ("strips=" + RamUsageTester.humanSizeOf(dic.stripData) + ", ")
         + ("conditions=" + RamUsageTester.humanSizeOf(dic.patterns) + ", ")
         + ("affixData=" + RamUsageTester.humanSizeOf(dic.affixData) + ", ")
+        + ("morphData=" + RamUsageTester.humanSizeOf(dic.morphData) + ", ")
```
Comment (donnerpeter) [inline lucene/analysis/common/src/test/org/apache/lucene/analysis/hunspell/TestAllDictionaries.java:208]: There's a slight memory usage increase, ~6% on average. I hope this is fine.

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/Dictionary.java [hunk 3]
```diff
@@ -1224,10 +1259,29 @@ public int compare(BytesRef o1, BytesRef o2) {
...
+      } else {
+        if (morphFields == null) morphFields = new ArrayList<>(1);
+        morphFields.add(datum);
+      }
+    }
+    return morphFields == null ? Collections.emptyList() : morphFields;
+  }
+
+  private int addMorphFields(Map<String, Integer> indices, String morphFields) {
+    Integer alreadyCached = indices.get(morphFields);
```
Comment (dweiss) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/Dictionary.java:1276]: indices.computeIfAbsent?

**Review by dweiss — COMMENTED** [review]
I left a note about one thing for you (maybe use computeIfAbsent?) but I'll merge this in - piggyback on another change, if you like.

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/Dictionary.java (hunk 3 above)
Comment (donnerpeter) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/Dictionary.java:1276]: Thanks, but my sense of beauty prefers `computeIfAbsent` lambdas to be side-effect free

Comment (dweiss) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/Dictionary.java:1276]: There is absolutely nothing stating they have to be though?... it's just a contract on updating the map - how it computes the value and what it updates along the way is not part of the contract at all.

I'll leave it to your sense of beauty though, of course.

Comment (donnerpeter) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/Dictionary.java:1276]: Yes, I've checked the javadoc and found nothing as well. But still, it aligns with other STM-like similar APIs (e.g. `AtomicInteger#getAndAccumulate`) and just feels better :)

**Pull request merged**